src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -438,11 +483,198 @@ StoragePtr DatabaseCatalog::tryGetTable(const StorageID & table_id) const
+        dot_pos = it.name().find('.', prev_dot_pos + 1);
+        if (dot_pos == std::string::npos)
+            continue;
+        dropped_id.uuid = parse<UUID>(it.name().substr(prev_dot_pos + 1, dot_pos - prev_dot_pos - 1));
+
+        String full_path = path + it.name();
+        dropped_metadata.emplace(std::move(full_path), std::move(dropped_id));
+    }
+
+    ThreadPool pool(SettingMaxThreads().getAutoValue());
+    for (const auto & elem : dropped_metadata)
+    {
+        pool.scheduleOrThrowOnError([&]()
+        {
+            this->enqueueDroppedTableCleanup(elem.second, nullptr, elem.first);
+        });
+    }
+    pool.wait();
+}
+
+String DatabaseCatalog::getPathForDroppedMetadata(const StorageID & table_id) const
+{
+    return global_context->getPath() + "metadata_dropped/" +
+           escapeForFileName(table_id.getDatabaseName()) + "." +
+           escapeForFileName(table_id.getTableName()) + "." +
+           toString(table_id.uuid) + ".sql";
+}
+
+void DatabaseCatalog::enqueueDroppedTableCleanup(StorageID table_id, StoragePtr table, String dropped_metadata_path, bool ignore_delay)
+{
+    assert(table_id.hasUUID());
+    assert(!table || table->getStorageID().uuid == table_id.uuid);
+    assert(dropped_metadata_path == getPathForDroppedMetadata(table_id));
+
+    time_t drop_time;
+    if (table)
+        drop_time = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    else
+    {
+        /// Try load table from metadata to drop it correctly (e.g. remove metadata from zk)
+        LOG_INFO(log, "Trying load partially dropped table " << table_id.getNameForLogs() << " from " << dropped_metadata_path);
+        ASTPtr ast = DatabaseOnDisk::parseQueryFromMetadata(log, *global_context, dropped_metadata_path, /*throw_on_error*/ false, /*remove_empty*/false);
+        auto create = typeid_cast<ASTCreateQuery *>(ast.get());
+        assert(!create || create->uuid == table_id.uuid);
+
+        if (create)
+        {
+            String data_path = "store/" + getPathForUUID(table_id.uuid);
+            create->database = table_id.database_name;
+            create->table = table_id.table_name;
+            try
+            {
+                table = createTableFromAST(*create, table_id.getDatabaseName(), data_path, *global_context, false).second;
+            }
+            catch (...)
+            {
+                tryLogCurrentException(log, "Cannot load partially dropped table " + table_id.getNameForLogs() +
+                                            " from: " + dropped_metadata_path +
+                                            ". Parsed query: " + serializeAST(*create) +
+                                            ". Will remove metadata and " + data_path +
+                                            ". Garbage may be left in ZooKeeper.");
+            }
+        }
+        else
+        {
+            LOG_WARNING(log, "Cannot parse metadata of partially dropped table "
```

Comment:
Seems like now it's possible just to execute `$ touch /var/lib/clickhouse/metadata_dropped/db_name.table_name.sql` and the table will be dropped after server restart. It's OK, but funny).

Reply:
No tables will be dropped, `db_name.table_name.sql` will be ignored because there is no UUID in file name and such files are not expected. If there is file like `db_name.table_name.283f4432-3c9f-4c31-bea3-01ad1ac9c749.sql` and it doesn't contain valid `CREATE` query, server will remove  this file and `/var/lib/clickhouse/store/283f4432-3c9f-4c31-bea3-01ad1ac9c749/` directory (if exists). I will add comments